internal/controller/apiextensions/composite/composed.go

```diff
@@ -54,6 +60,224 @@ const (
+type TemplateAssociation struct {
+	Template  v1.ComposedTemplate
+	Reference corev1.ObjectReference
+}
+
+// AssociateByOrder associates the supplied templates with the supplied resource
+// references by order; i.e. by assuming template n corresponds to reference n.
+// The returned array will always be of the same length as the supplied array of
+// templates. Any additional references will be truncated.
+func AssociateByOrder(t []v1.ComposedTemplate, r []corev1.ObjectReference) []TemplateAssociation {
+	a := make([]TemplateAssociation, len(t))
+	for i := range t {
+		a[i] = TemplateAssociation{Template: t[i]}
+	}
+
+	j := len(t)
+	if len(r) < j {
+		j = len(r)
+	}
+
+	for i := 0; i < j; i++ {
+		a[i].Reference = r[i]
+	}
+
+	return a
+}
+
+// A CompositionTemplateAssociator returns an array of template associations.
+type CompositionTemplateAssociator interface {
+	AssociateTemplates(context.Context, resource.Composite, *v1.Composition) ([]TemplateAssociation, error)
+}
+
+// A CompositionTemplateAssociatorFn returns an array of template associations.
+type CompositionTemplateAssociatorFn func(context.Context, resource.Composite, *v1.Composition) ([]TemplateAssociation, error)
+
+// AssociateTemplates with composed resources.
+func (fn CompositionTemplateAssociatorFn) AssociateTemplates(ctx context.Context, cr resource.Composite, comp *v1.Composition) ([]TemplateAssociation, error) {
+	return fn(ctx, cr, comp)
+}
+
+// A GarbageCollectingAssociator associates a Composition's resource templates
+// with (references to) composed resources. It tries to associate them by
+// checking the template name annotation of each referenced resource. If any
+// template or existing composed resource can't be associated by name it falls
+// back to associating them by order. If it encounters a referenced resource
+// that corresponds to a non-extistent template the resource will be garbage
+// collected (i.e. deleted).
+type GarbageCollectingAssociator struct {
+	client client.Client
+}
+
+// NewGarbageCollectingAssociator returns a CompositionTemplateAssociator that
+// may garbage collect composed resources.
+func NewGarbageCollectingAssociator(c client.Client) *GarbageCollectingAssociator {
+	return &GarbageCollectingAssociator{client: c}
+}
+
+// AssociateTemplates with composed resources.
+func (a *GarbageCollectingAssociator) AssociateTemplates(ctx context.Context, cr resource.Composite, comp *v1.Composition) ([]TemplateAssociation, error) { //nolint:gocyclo
+	// NOTE(negz): This method is a little over our complexity goal. Be wary of
+	// making it more complex.
+
+	templates := map[string]int{}
+	for i, t := range comp.Spec.Resources {
+		if t.TemplateName == nil {
+			// If our templates aren't named we fall back to assuming that the
+			// existing resource reference array (if any) already matches the
+			// order of our resource template array.
+			return AssociateByOrder(comp.Spec.Resources, cr.GetResourceReferences()), nil
```

Comment:
It would, but technically doing so would result in a behaviour change when folks upgraded to Crossplane 1.1 (presuming this ships with 1.1). Currently there's no behaviour change unless they explicitly update their Compositions to include template/base names.

I'd prefer to go with the simpler and more conservative approach here; I think going forward we should strongly recommend folks use template/base names so anyone who does will get garbage collection support.